### PR TITLE
Csv parse

### PR DESCRIPTION
**Pull request by amrsoll**
Parse a csv file in order to change the material settings served in the front end

### CODE REVIEW

File: octoprint_mrbeam/__init__.py [hunk 1]
```diff
@@ -1254,6 +1255,10 @@ def on_api_command(self, command, data):
 			return self.usage_handler.reset_laser_head_usage()
 		elif command == "reset_gantry_usage":
 			return self.usage_handler.reset_gantry_usage()
+		elif command == "material_settings":
+			# TODO select which Mr Beam version to parse the materials for
+			# TODO Load materials when the user logs in too
+			return make_response(jsonify(parse_csv()), 200)
```
Comment (amrsoll) [inline octoprint_mrbeam/__init__.py]: return make_response("err", 500) if failed to parse.

Comment (amrsoll) [reply octoprint_mrbeam/__init__.py]: + raise exceptions in parse_csv

Comment (amrsoll) [reply octoprint_mrbeam/__init__.py]: self.logger.exception(...)

File: octoprint_mrbeam/static/js/material_settings.js [hunk 2]
```diff
@@ -5,23 +5,37 @@ $(function () {
...
+                        }
                     }
-                    self.materialSettingsDatabase[materialKey].laser_type = my_laser_type;
-                    self.materialSettingsDatabase[materialKey].colors = self.materialSettingsDatabase[materialKey].params[my_laser_type];
-                }
-            }
+                    console.log("loadMaterials callback: ", self.materialSettingsDatabase);
+                    callback(self.materialSettingsDatabase);
+                });
+
```
Comment (amrsoll) [inline octoprint_mrbeam/static/js/material_settings.js]: .fail( ...) -> handle exception from the result

Comment (amrsoll) [reply octoprint_mrbeam/static/js/material_settings.js]: log failed to load material settings

File: octoprint_mrbeam/static/js/material_settings.js [hunk 3]
```diff
@@ -5,23 +5,37 @@ $(function () {
...
-
         self.default_laser_type = 'MrBeamII-1.0';
 
-        self.getMaterialSettings = function (laser_type) {
-            for (let materialKey in self.materialSettingsDatabase) {
-                if (self.materialSettingsDatabase[materialKey] && self.materialSettingsDatabase[materialKey].params) {
-                    let my_laser_type = laser_type || self.default_laser_type
-                    if (!(my_laser_type in self.materialSettingsDatabase[materialKey].params)) {
-                        my_laser_type = self.default_laser_type;
+        self.getMaterialSettings = function (callback) {
```
Comment (andywerner) [inline octoprint_mrbeam/static/js/material_settings.js]: rename to loadMaterialSettings

File: octoprint_mrbeam/static/js/material_settings.js [hunk 4]
```diff
@@ -5,23 +5,37 @@ $(function () {
...
-                }
-            }
+                    console.log("loadMaterials callback: ", self.materialSettingsDatabase);
+                    callback(self.materialSettingsDatabase);
+                });
+
             return self.materialSettingsDatabase;
         };
 
+        self.loadMaterials = function (callback) {
```
Comment (andywerner) [inline octoprint_mrbeam/static/js/material_settings.js]: remove this funciton

File: octoprint_mrbeam/static/js/material_settings.js [hunk 5]
```diff
@@ -32,46 +46,30 @@ $(function () {
                 img: 'anodized_aluminum.jpg',
                 description: gettext("Dark anodized aluminum can be engraved. Works on iPhones."),
                 hints: gettext("Requires very precise focus. Anodized aluminum turns brighter through laser engraving. Therefore we suggest to invert photos for engravings."),
-                params: {
-                    'MrBeamII-1.0': {
-                        '000000': {
-                            engrave: {eng_i: [0, 100], eng_f: [1000, 30], eng_pierce: 0, dithering: false},
+                colors: {
```
Comment (andywerner) [inline octoprint_mrbeam/static/js/material_settings.js]: remove all the colors values but keep one as a comment for future developer's reference

File: octoprint_mrbeam/static/js/material_settings.js [hunk 6]
```diff
@@ -32,46 +46,30 @@ $(function () {
...
+                    'd4b26f': {
+                        engrave: {eng_i: [0, 20], eng_f: [2000, 350], engrave_compressor_lvl: 0, eng_pierce: 0, dithering: false},
+                        cut: [
+                            {thicknessMM: 1, cut_i: 80, cut_f: 525, cut_p: 2},
+                            {thicknessMM: 2, cut_i: 100, cut_f: 525, cut_p: 2},
+                            {thicknessMM: 3, cut_i: 100, cut_f: 525, cut_p: 2},
+                            {thicknessMM: 4, cut_i: 100, cut_f: 450, cut_p: 3},
+                            {thicknessMM: 5, cut_i: 100, cut_f: 225, cut_p: 3}
+                        ]
+                    }
```
Comment (amrsoll) [inline octoprint_mrbeam/static/js/material_settings.js]: Delete all previous material settings, keep one in comments as an example of the data structure

**Review by andywerner — CHANGES_REQUESTED** [review]

File: octoprint_mrbeam/util/material_csv_parser.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
+import sys, os, csv, json, collections
+
+MRBEAM = 'Mr Beam II'
+DREAMCUT = 'Dreamcut'
+READY = 'Dreamcut Ready'
```
Comment (amrsoll) [inline octoprint_mrbeam/util/material_csv_parser.py]: Shorten to MRB but add it to dreamcut and dreamcut ready

File: octoprint_mrbeam/util/material_csv_parser.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                    material = prev_vals[1]
+                if not colorcode:  # No Color name, assume the previous one
+                    colorcode = prev_vals[2]
+
+            try:
+                thickness = float(thickness_or_engrave)
+                settingname = 'cut'
+                settings = [{'thicknessMM': thickness, 'cut_i': int(intensity), 'cut_f': int(speed), 'cut_p': int(passes), 'cut_compressor_lvl': int(compressor_lvl)}]
+            except ValueError:
+                dithering = True if dithering == 'yes' else False
```
Comment (amrsoll) [inline octoprint_mrbeam/util/material_csv_parser.py:79]: Test if the string is "engrave"-ish
and raise an exception with linecounter

File: octoprint_mrbeam/util/material_csv_parser.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+                pierce_time = pierce_time or 0
+                settings = {'engrave_compressor_lvl': int(compressor_lvl),
+							'eng_pierce': int(pierce_time),
+                            'dithering': dithering}
+                settingname = 'engrave'
+                i_split = intensity.split('-')
+                if len(i_split) == 2:
```
Comment (amrsoll) [inline octoprint_mrbeam/util/material_csv_parser.py:86]: add string.strip() as failsafe

**Review by amrsoll — COMMENTED** [review]
pretty noice

**Pull request merged**